pubsub/gossipsub/partial-messages.md

```diff
@@ -0,0 +1,189 @@
+be referenced by bitmaps, ranges, or a bloom filter.
+
+The structure of partial messages in a group, as well as how partial messages
+are referenced is application defined.
+
+If, in some application, a group only ever contained a single partial message,
+then partial messages would be the same as smaller messages.
+
+
+## Protocol Messages
+
+The following section specifies the semantics of each new protocol message.
+
+### PartialIWANT
+
+Partial IWants signal to a receiver that the sending peer only wants a part of
+some message.
+
+The message to which the peer is requesting a part of is identified by the
+`groupID` identifier. This is similar to a complete message's `MessageID`, but,
+in contrast to a content-based message id, does not require the full message to
+compute. For example, in the Ethereum use case, this could simply be the hash of
+the signed block header.
+
+The `topicID` references the Gossipsub topic a message, and thus its parts,
+belong to.
+
+The `metadata` field is opaque application defined metadata associated with this
+request. This can be a bitmap, a list of ranges, or a bloom filter. The
+application generates this and consumes this.
+
+
+Nodes SHOULD assume a `partialIWANT `implies a `IDONTWANT `for the full message.
+
+### PartialIDONTWANT
+
+PartialIDONTWANT serves to cancel any pending PartialIWANTs.
+
+A PartialIDONTWANT without metadata serves to cancel all pending PartialIWANTs,
+and signals there is no part this peer wants.
+
+Implementations SHOULD NOT send a PartialIHAVE to a peer with parts that the
+peer has previously sent a PartialIDONTWANT for.
+
+### PartialIHAVE
+
+PartialIHave allows nodes to signal HAVE information before receiving all
+segments, unlocking the use of partialIWANT in more contexts.
+
+In the context of partial messages, it is more useful than IHAVE as it tells the
+peer the group ID. In contrast, a IHAVE only includes a message ID that is
+unique to the message. A receiving peer has no way to link an IHAVE's message ID
+with a group ID, without having the full message.
+
+Partial IHAVE messages can be used both in the context of lazy push, notifying
+peers about reception progress, and in the context of heartbeats, sending
+also Partial IHAVEs.
+
+The structure of PartialIHAVE is analogous to that of PartialIWANT.
+
+The metadata, as in Partial IWants, is application defined. It is some encoding
+that represents the parts the sender has.
+
+Implementations are free to select when to send an update to their peers based
+on signaling bandwidth tradeoff considerations.
+
+## Application Interface
+
+Message contents are application defined. Thus splitting a message must be
+application defined. Here are a list of operations an application is expected to
+provide to Gossipsub to enable partial message delivery.
+
+1. Split a full message into a group of partial messages.
+2. Given a group of partial messages, encode a request for the missing parts. (Used by `PartialIWANT`)
```

Comment:
This will require some higher level contextual information also, right? In our case, in addition to the actual partial messages generated in (1), we need to know the total number of parts and the positions our local parts occupy within the series, so we can generate the bitmap.